File: crates/meta-extract/src/a1111.rs
use crate::models::{MetaResult, PngTextChunks};
use std::collections::HashMap;

pub fn parse_a1111(chunks: &PngTextChunks) -> Option<MetaResult> {
    let raw = chunks
        .entries
        .get("parameters")
        .or_else(|| chunks.entries.get("Parameters"))
        .or_else(|| chunks.entries.get("PARAMETERS"))?
        .clone();
    Some(parse_a1111_text(&raw))
}

pub fn parse_a1111_text(raw: &str) -> MetaResult {
    let text = raw.trim_start_matches('\u{feff}').replace("\r\n", "\n");
    let text = text
        .strip_prefix("Parameters:")
        .unwrap_or(&text)
        .trim_start()
        .to_string();

    // Normalize mid-line section markers to newline-prefixed form.
    let text = text.replace(". Negative prompt:", "\nNegative prompt:");
    let text = text.replace(". Steps:", "\nSteps:");

    let neg_marker = "\nNegative prompt:";
    let steps_marker = "\nSteps:";

    // Split off positive
    let (positive, rest) = if let Some(pos) = text.find(neg_marker) {
        (text[..pos].to_string(), text[pos + 1..].to_string())
    } else if let Some(pos) = text.find(steps_marker) {
        (text[..pos].to_string(), text[pos + 1..].to_string())
    } else {
        return MetaResult {
            positive: Some(text.trim().to_string()),
            format: "a1111".into(),
            raw_meta: Some(raw.to_string()),
            ..Default::default()
        };
    };

    // Split rest into negative / params
    let (negative_opt, params_str) = if let Some(after) = rest.strip_prefix("Negative prompt:") {
        if let Some(pos) = after.find("\nSteps:") {
            (
                Some(after[..pos].trim().to_string()),
                Some(after[pos + 1..].to_string()),
            )
        } else {
            (Some(after.trim().to_string()), None)
        }
    } else if rest.starts_with("Steps:") {
        (None, Some(rest.clone()))
    } else {
        (None, None)
    };

    let params = params_str
        .as_deref()
        .map(parse_params_section)
        .unwrap_or_default();

    MetaResult {
        positive: Some(positive.trim().to_string()),
        negative: negative_opt,
        format: "a1111".into(),
        raw_meta: Some(raw.to_string()),
        params,
    }
}

fn parse_params_section(s: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    let line = s.lines().next().unwrap_or(s);
    for part in line.split(',') {
        let part = part.trim();
        if let Some(colon) = part.find(':') {
            let key = part[..colon].trim().to_string();
            let val = part[colon + 1..].trim().to_string();
            if !key.is_empty() {
                map.insert(key, val);
            }
        }
    }
    map
}

#[cfg(test)]
mod tests {
    use super::*;

    fn chunks(params: &str) -> PngTextChunks {
        let mut c = PngTextChunks::default();
        c.entries.insert("parameters".into(), params.into());
        c
    }

    #[test]
    fn basic_pos_neg_params() {
        let raw = "masterpiece, best quality\nNegative prompt: lowres, bad anatomy\nSteps: 20, Sampler: Euler a, CFG scale: 7";
        let r = parse_a1111(&chunks(raw)).unwrap();
        assert_eq!(r.positive.as_deref(), Some("masterpiece, best quality"));
        assert_eq!(r.negative.as_deref(), Some("lowres, bad anatomy"));
        assert_eq!(r.params.get("Steps").map(|s| s.as_str()), Some("20"));
        assert_eq!(r.format, "a1111");
    }

    #[test]
    fn no_negative() {
        let raw = "a cat\nSteps: 10, Sampler: DPM";
        let r = parse_a1111(&chunks(raw)).unwrap();
        assert_eq!(r.positive.as_deref(), Some("a cat"));
        assert!(r.negative.is_none());
    }

    #[test]
    fn no_steps() {
        let raw = "a dog";
        let r = parse_a1111(&chunks(raw)).unwrap();
        assert_eq!(r.positive.as_deref(), Some("a dog"));
        assert!(r.params.is_empty());
    }

    #[test]
    fn parameters_prefix_stripped() {
        let raw = "Parameters: a fox\nSteps: 5";
        let r = parse_a1111(&chunks(raw)).unwrap();
        assert_eq!(r.positive.as_deref(), Some("a fox"));
    }
}

File: core/tagdb_prompt/a1111.py
import re

from .models import A1111Parsed
from .utils import norm_space


def parse_a1111_parameters(params_text: str) -> A1111Parsed:
    text = (params_text or "").strip()
    text = text.lstrip(chr(65279))
    if not text:
        return A1111Parsed("", "", {})

    text = re.sub(r"^\s*Parameters?\s*:\s*", "", text, flags=re.IGNORECASE)
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = re.sub(r"\.\s*(Negative prompt:)", r"\n\1", text, flags=re.IGNORECASE)
    text = re.sub(r"\.\s*(Steps:)", r"\n\1", text, flags=re.IGNORECASE)
    text = re.sub(r"\s+(Negative prompt:)", r"\n\1", text, flags=re.IGNORECASE)
    text = re.sub(r"\s+(Steps:)", r"\n\1", text, flags=re.IGNORECASE)
    lines = text.splitlines()

    pos_lines: list[str] = []
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        if re.match(r"^Negative prompt:", line, flags=re.IGNORECASE) or line.startswith("Steps:"):
            break
        pos_lines.append(lines[i])
        i += 1
    positive = "\n".join(pos_lines).strip()

    neg_lines: list[str] = []
    if i < len(lines) and re.match(r"^Negative prompt:", lines[i].strip(), flags=re.IGNORECASE):
        first = re.sub(r"^Negative prompt:\s*", "", lines[i].strip(), flags=re.IGNORECASE)
        if first:
            neg_lines.append(first)
        i += 1
        while i < len(lines) and not lines[i].strip().startswith("Steps:"):
            if norm_space(lines[i]):
                neg_lines.append(lines[i].strip())
            i += 1
    negative = "\n".join(neg_lines).strip()

    params: dict[str, str] = {}
    if i < len(lines):
        param_line = lines[i].strip()
        if param_line.startswith("Steps:"):
            parts = [p.strip() for p in param_line.split(",") if p.strip()]
            for part in parts:
                if ":" not in part:
                    continue
                k, v = part.split(":", 1)
                k = norm_space(k)
                v = norm_space(v)
                if k:
                    params[k] = v
    return A1111Parsed(positive, negative, params)
